### PR TITLE
Version 1.0

### PR DESCRIPTION
**Pull request by bgiori**
**Brief Notes**

* Rather than changing current modules, create two new modules
    - `core`: Contains API and transport interface definition. Also contains the new, faster window upload and download.
    - `serialization`: Contains encoder, decoder, CBOR object mapper and base message type.
* Transport has been stripped of connection stuff. My idea is that DFU will use a separate connection interface in order to not pollute the transport interface.
* Transport has been changed so send and receive objects rather than send bytes and receive objects. 
    - I.e. transport implementation will need to perform the encoding and decoding.
* Request & Response objects are implemented with sealed classes which enforce nullability.
    - API is much more well defined, and it should be easy to spot inconsistencies.
* Transport and initial API are all suspend functions. Library should be much less blocking.

More documentation later.

### CODE REVIEW

**Comment by bgiori** [comment]
@philips77 Let me know what you think of this design. No need for a full code review; just wanted to get your input on the restructuring.

**Comment by philips77** [comment]
I like the new structure. Also, I went quickly through the code and looks very good. Eager to see the end result. Do you plan to make the API backwards compatible, or as it's 1.0 you don't have to carry this baggage? And, are you planning on adjusting the sample app when libraries are complete?

**Comment by bgiori** [comment]
I think backwards compatibility might be off the table due to the major type changes (e.g. McuMgrResponse, McuManager, etc.) and the package name changes. However, mimicking the previous callback style APIs to lessen the effect of updating from 0.x to 1.x should be easy.

> And, are you planning on adjusting the sample app when libraries are complete?

Yes, I am. Although I think that will be the last item on the TODO list.

File: core/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (twyatt) [inline core/.gitignore:1]: nitpick

If you have `build/` in your root `.gitignore` then you don't have to have a `.gitignore` for each module.

Related discussion: https://github.com/JuulLabs/juul-app-android/pull/1992#discussion_r406665454

_(screenshot for those without access):_

Screen Shot 2020-07-09 at 3 04 15 PM

File: core/src/main/kotlin/McuMgrResult.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+inline fun McuMgrResult<*>.onErrorOrFailure(action: (throwable: Throwable) -> Unit): McuMgrResult<*> {
+    when (this) {
+        is Error -> action(ErrorCodeException(code))
+        is Failure -> action(throwable)
+    }
+    return this
+}
+
+// Get Or
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:63]: nitpick: superfluous comment

File: core/src/main/kotlin/McuMgrResult.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        get() = this is Failure
+
+    fun exceptionOrNull(): Throwable? =
+        when (this) {
+            is Failure -> throwable
+            else -> null
+        }
+}
+
+// On Blocks
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:32]: nitpick: superfluous comment

File: core/src/main/kotlin/McuMgrResult.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+inline fun <T> McuMgrResult<T>.getOrElse(action: (exception: Throwable) -> T): T {
+    return when (this) {
+        is Success -> value
+        is Error -> action(ErrorCodeException(code))
+        is Failure -> action(throwable)
+    }
+}
+
+// Transformations
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:81]: nitpick: superfluous comment

File: core/src/main/kotlin/McuMgrResult.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+inline fun <R, T> McuMgrResult<T>.map(transform: (value: T) -> R): McuMgrResult<R> {
+    return when (this) {
+        is Success -> Success(transform(value))
+        is Error -> Error(code)
+        is Failure -> Failure(throwable)
+    }
+}
+
+// Exception
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:91]: nitpick: superfluous comment

File: core/src/main/kotlin/McuMgrResult.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+class ErrorCodeException(
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:93]:
```suggestion
class ErrorCodeException internal constructor(
```

File: core/src/main/kotlin/McuMgrResult.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    val code: Response.Code
+) : Exception("Request resulted in error response $code")
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:95]: nitpick: Generally better to pick a more specific exception.

Perhaps `IOException` or `IllegalStateException`?

File: core/src/main/kotlin/message/Message.kt [hunk 8]
```diff
@@ -0,0 +1,132 @@
+package com.juul.mcumgr.message
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.annotation.JsonValue
```
Comment (twyatt) [inline core/src/main/kotlin/message/Message.kt:6]: Ideally, you'd decouple your serialization from your public API.

Essentially, it's be best to expose these data structures without the serialization annotations, then you have a "bridge" layer that copies data from your public API structure to your serialization specific objects.

This allows your serialization tool and data structure to change without breaking the public facing API. Also means your public API isn't directly driven by the underlying serialization structure. So you can expose more Kotlin idiomatic data structures than maybe the underlying firmware adheres to.

File: serialization/src/main/kotlin/Encoder.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+private fun Message.encodeStandard(): ByteArray =
+    Buffer().apply {
+        // Set the header's length to the size of the payload.
+        val payloadBytes = encodePayload()
+        val lengthSetHeader = header.copy(length = payloadBytes.size)
+        encodeHeader(lengthSetHeader)
+        write(payloadBytes)
+    }.readByteArray()
+
+private fun Message.encodeCoap(): ByteArray {
```
Comment (twyatt) [inline serialization/src/main/kotlin/Encoder.kt:38]: Is this encoding CoAP or CBOR?

File: core/src/main/kotlin/transfer/Uploader.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    val format: Format
+
+    suspend fun write(data: ByteArray, offset: Int, length: Int): McuMgrResult<Response>
+
+    data class Event(val offset: Int, val length: Int)
+    data class Response(val offset: Int)
+}
+
+@Throws
+suspend fun Uploader.flow(
```
Comment (twyatt) [inline core/src/main/kotlin/transfer/Uploader.kt:26]: I don't quite follow why this is a `Flow`?
Perhaps I'm just having trouble understanding what this function is doing.

Is it grabbing chunks of data that will ultimately be uploaded?

File: core/src/main/kotlin/transfer/Downloader.kt [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * Validate the the data size is as expected.
+ *
+ * Handles the case of the last data packet usually being smaller than the expected size.
+ */
+private fun assertDataSize(actualSize: Int, expectedSize: Int, offset: Int, totalSize: Int) {
+    if (actualSize != expectedSize && offset + actualSize != totalSize) {
+        throw IllegalStateException("data size does not match expected size")
+    }
```
Comment (twyatt) [inline core/src/main/kotlin/transfer/Downloader.kt:100]: nitpick

You can use the more idiomatic [`check`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/check.html).

File: core/src/main/kotlin/transfer/Downloader.kt [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    suspend fun read(offset: Int): McuMgrResult<Response>
+
+    data class Event(val data: ByteArray, val offset: Int, val length: Int)
+    data class Response(val data: ByteArray, val offset: Int, val length: Int?)
+}
+
+@Throws
+fun Downloader.flow(
+    capacity: Int
+) = channelFlow {
```
Comment (twyatt) [inline core/src/main/kotlin/transfer/Downloader.kt:26]: I think I kind of understand what's going on here, but if so, then it doesn't seem you need a `Flow`.
`Flow` is useful when you want to provide a lazy on-demand stream of data, but from what I can tell, you'll explicitly want to perform the upload. i.e. Standard Coroutines will accomplish what you want and me a lower cognitive lift to understand the code.

File: serialization/src/main/kotlin/Encoder.kt (hunk 9 above)
Comment (bgiori) [reply serialization/src/main/kotlin/Encoder.kt:38]: Same as the other comment, this is the "Coap" mcumgr message format for CoAP based transports.  

The other name would be SMP vs OMP, which is probably better now that I think about it. as that is what the other mcumgr libraries use.

File: core/src/main/kotlin/message/Message.kt (hunk 8 above)
Comment (bgiori) [reply core/src/main/kotlin/message/Message.kt:6]: I think I understand. Since we are using jackson I would want to create custom serializers for each request and response & add them to the ObjectMapper? That way the objects do not need to be annotated. Akin to a moshi adapter?

File: serialization/src/main/kotlin/Decoder.kt [hunk 13]
```diff
@@ -1,57 +1,54 @@
...
-        val headerBuffer = Buffer().write(rawHeader)
-        val header = headerBuffer.decodeHeader()
-        // Decode the whole payload as the message type
-        val buffer = Buffer().write(data)
-        val payload = buffer.decodePayload(type)
-        Message(header, payload)
-    } catch (e: IOException) {
-        throw DecoderException(e.message ?: "Failed to decode message.", e, data, type.name)
-    }
+inline fun <reified R> ByteArray.decodeCoap(): Message.Coap<R> = try {
```
[the remark this answers is not in the document]
Comment (bgiori) [reply serialization/src/main/kotlin/Decoder.kt]: Coap in this case is the message format. mcumgr uses a separate format for coap based transport implementations.

File: core/src/main/kotlin/McuMgrResult.kt (hunk 6 above)
Comment (bgiori) [reply core/src/main/kotlin/McuMgrResult.kt:93]: public inline functions cannot access internal api

File: core/src/main/kotlin/message/Message.kt (hunk 8 above)
Comment (twyatt) [reply core/src/main/kotlin/message/Message.kt:6]: Doing the conversion within adapters works. More importantly though it's about what is exposed on the public API surface.

Essentially you want something like:

```kotlin
data class PublicApiObject(
    val a: Foo,
    val b: Bar
)

internal SerializationObject(
    val y: Hello,
    val z: World
)
```

The `SerializationObject` can be annotated with serialization annotations if the serialization library uses them.
How you get from `PublicApiObject` to `SerializationObject` could either be manually or via serialization library adapters.

The ultimate goal being decoupling, whereas here, there structure of `PublicApiObject` can be whatever makes the most sense for a library consumer (and can be idiomatic to Kotlin types). It also can remain unchanged even if the underlying serialization structure changes; as the library would just update it's conversion internals if changes occurred.

Comment (bgiori) [reply core/src/main/kotlin/message/Message.kt]: So I fiddled with custom serializers (adapters) and felt it was overly complex and did not "describe" the API like an annotated data class. 

Instead I used a jackson feature called a mixin. Basically substituting a non-annotated class with an annotated one. You then add the mapping between the two classes to the object mapper, and viola, it works.

The new serializations objects I called payloads and put in the serialization module under `Payload.kt`.

File: core/src/main/kotlin/transfer/Uploader.kt (hunk 10 above)
Comment (bgiori) [reply core/src/main/kotlin/transfer/Uploader.kt:26]: Resolving to move discussion to other comment

File: core/src/main/kotlin/transfer/Downloader.kt (hunk 12 above)
Comment (bgiori) [reply core/src/main/kotlin/transfer/Downloader.kt:26]: Sounds good. I think will implement this with a returned receive channel instead.

File: core/src/main/kotlin/message/Message.kt (hunk 8 above)
Comment (twyatt) [reply core/src/main/kotlin/message/Message.kt]: Great! Really, whatever gets the job done to convert from your public API to your private serialization API (and allows your public API to have different structure / variable naming than your serialization API — if needed). 👍

File: core/src/main/kotlin/transfer/Downloader.kt (hunk 12 above)
Comment (twyatt) [reply core/src/main/kotlin/transfer/Downloader.kt:26]: Not sure returning a `Channel` is the right approach either. Either that or I'm very much misunderstanding the purpose of this function.

Don't you just want to call a function and perform some downloads?
Why does that need to be "subscription" based? Why can't it be a `suspend` function (or `CoroutineScope` extension function)?

Unless this is an ongoing data stream? Whereas downloads could "appear" at any given time, and the `Flow` allows a library consumer to subscribe and be informed of these sporadic download events?

Comment (bgiori) [reply core/src/main/kotlin/transfer/Downloader.kt:26]: I'm not set on using a Flow by any means, however I'd like the transfer to be a stream for progress callbacks which can drive a UI element (e.g. image upload used in DFU).

**Comment by bgiori** [comment]
> Does serialization actually need to be a separate module? Will other modules need to use that code (or is it specific to core)?

@twyatt Whatever implements `Transport` will need both `core` and `serialization`, but the user of the transport only requires what is in `core` and `transport`, which is why serialization is modularized. Most users of this library use the premade transport, and thus do not need what is in serialization. One of the main complexities of the current implementation 0.x is the `McuMgrResponse` object which handles both the SMP and OMP formats and exposes way more than is needed to the user.

To sum up, most users don't need access to anything in `serialization` unless they are implementing their own transport layer.

**Comment by twyatt** [comment]
> > Does serialization actually need to be a separate module? Will other modules need to use that code (or is it specific to core)?
> 
> @twyatt Whatever implements `Transport` will need both `core` and `serialization`, but the user of the transport only requires what is in `core` and `transport`, which is why serialization is modularized. Most users of this library use the premade transport, and thus do not need what is in serialization. One of the main complexities of the current implementation 0.x is the `McuMgrResponse` object which handles both the SMP and OMP formats and exposes way more than is needed to the user.
> 
> To sum up, most users don't need access to anything in `serialization` unless they are implementing their own transport layer.

Got it. Makes sense if developers extending functionality of the library need access to the serialization to have it as a separate module. Thanks for clarifying. 👍

File: core/src/main/kotlin/transfer/Downloader.kt (hunk 12 above)
Comment (twyatt) [reply core/src/main/kotlin/transfer/Downloader.kt:26]: Oh gotchya. For the purposes of propagating status/state a data stream does make sense.

What about this for an API?:

```kotlin
class Downloader(private resource: URI) {
    val progress: Flow<Int> // hot conflated Flow that emits values from 0-100
    suspend fun download(): ByteArray
}
```

Here, a call to `download` will suspend until complete. If developer wants status of the download (i.e. for UI updating purposes) then they can optionally subscribe to `progress`.

You can definitely make `progress` more fancy, with a `sealed class` that represents things like `InProgress(percent: Float)`, `Complete` or `Failure`.

Subscribing or unsubscribing to `progress` has no effect on the download (so is fine to subscribe from, for example, a `LiveData` that may unsubscribe and re-subscribe based on lifecycle; i.e. device rotation).

You `download` function can throw an exception on failure, or return a domain-specific `Result` object; up to you.

Comment (twyatt) [reply core/src/main/kotlin/transfer/Downloader.kt:26]: You can get ultra fancy with something like:

```kotlin
suspend fun download(
    resource: URI,
    progressHandler: ((Int) -> Unit)? = null
): ByteArray {
    val downloader = Downloader(resource)
    return coroutineScope { 
        val job = downloader.progress.onEach { progressHandler?.invoke(it) }.launchIn(this)
        val result = downloader.download()
        progressHandler?.invoke(100)
        job.cancel()
        result
    }
}
```

It's usage could like:

```kotlin
val bytes = download(uri) { progress -> /* todo: update UI */ }
```

**Comment by bgiori** [comment]
@twyatt I have updated the upload and download classes with a new API, added tests, and fixed a few bugs. 

Notes:
 * Upload and Download are now abstract classes rather than interfaces.
 * Upload and Download now contain member `progress` for progress updates
 * Moved capacity variable to constructor.
 * User friendly APIs have been created in `McuManager`

Image upload is used as an example of API

```kotlin
suspend fun uploadImage(
    data: ByteArray,
    windowCapacity: Int = 1,
    progressHandler: ((UploadProgress) -> Unit)? = null
): McuMgrResult<Unit>
```

Example Usage

```kotlin
uploadImage(imageData, 5) {
    // Update progress UI
}.onErrorOrFailure {
   // Handle failure
}
```
---------------------

On a side note, I noticed in your more recent projects that the way gradle dependencies are handled has changed. E.g. dependencies are now listed in `gradle/dependencies.gradle`. Is this the new hot way to import gradle dependencies? Should it be something to follow suit?

**Comment by twyatt** [comment]
> Is this the new hot way to import gradle dependencies? Should it be something to follow suit?

Not sure if it's the "new hot way" but it's just something I adopted to organize things a bit more.
I think the new hotness is `build.gradle.kts` (using Kotlin as the Gradle language rather than Groovy) which I haven't yet had the time to catch up on.

The `gradle/dependencies.gradle` just seemed like a easy way to put all your dependency definitions in one place and made version management much easier (IMHO).

So you're definitely encouraged to follow the approach, or any other approach that improves the ease of configuring the project. :)

File: core/src/main/kotlin/transfer/Uploader.kt [hunk 14]
```diff
@@ -1,133 +1,185 @@
...
+    suspend fun upload() = coroutineScope {
+
+        var offset = 0
+
+        while (offset < data.size) {
+
+            val transmitOffset = offset
+            val chunkSize = getChunkSize(data, offset)
+
+            window.acquire()
```
[the remark this answers is not in the document]
Comment (bgiori) [reply core/src/main/kotlin/transfer/Uploader.kt:50]: Not for serialization, buf for dynamic flow control. We only want to have up to `capacity` number of active transactions (requests/responses) at a time in order to not overload the device into OOM. Further, if we receive an error for a request, we want to reduce the window size.

`WindowSemaphore` is the class which encapsulates this logic. In short, the window is initialized with a `capacity` which is the _max_ `size` the window can reach. The `size` starts at 1 and grows by 1 with every successful request until either (a) the capacity is reached, or (b) a request errors out. On error (b) the window shrinks by one and a flag is set such that the window will no longer grow.

Example window `capacity=4`
* `-->` == send request
* `<--` == receive response

```
--> id=1
<-- id=1, result=success (size = 2)
--> id=2
--> id=3
<-- id=2, result=success (size = 3)
--> id=4
--> id=5
<-- id=3 result=success (size = 4)
--> id=6
--> id=7
<-- id=4 result=success (size = 4) // Reached capacity
--> id=8
<-- id=5 result=success (size = 4)
--> id=9
<-- id=6 result = FAIL (size = 3) // Shrink and retry
--> id=6
<-- id=7 result = success (size = 3) // id=6 error means no more growth
<-- id=8 result = success (size = 3)
--> id=10
<-- id=9 result = success (size = 3)
--> id=11
<-- id=6 result = success (size = 3) // recover from error, release permit

...
```

While I was writing up this timeline I noticed a bug in the current implementation which I justl fixed.

File: core/src/main/kotlin/McuManager.kt [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        val uploader = FileUploader(data, fileName, this, windowCapacity)
+        return catchResult {
+            upload(uploader, progressHandler)
+        }
+    }
+
+    @Throws
+    suspend fun upload(
+        uploader: Uploader,
+        progressHandler: ((UploadProgress) -> Unit)? = null
```
Comment (twyatt) [inline core/src/main/kotlin/McuManager.kt:89]: This feels like Java callback style when you have Coroutines to provide data streams with all its fancy operators.

In fact, it looks like you're wrapping a nice Coroutines API in this callback style function.

This function feels like a good candidate for a separate module where you bridge the Coroutine and Java world. Doesn't feel like it belongs in this module.

Instead, you could have:

```kotlin
class McuManager {
    // ...

    fun fileUploader(
        data: ByteArray,
        fileName: String,
        windowCapacity: Int = 1
    ) = FileUploader(data, fileName, this, windowCapacity)

    // ...
}
```

Users of library can easily call and use it as they see fit:

```kotlin
val mcumgr: McuManager

// When progress isn't needed.
mcumgr.fileUploader(...).upload() // Suspends until complete.
```

If progress is desired then one possible usage could look like:

```kotlin
val mcumgr: McuManager
val uploader = mcumgr.fileUploader(...)
uploader
    .progress
    .transformWhile { progress ->
        emit(progress)
        progress != 100
    }
    .onEach { ... }
    .launchIn(scope)
uploader.upload() // Suspends until complete.
```

File: core/src/main/kotlin/McuMgrResult.kt [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+sealed class McuMgrResult<out T> {
+
+    data class Success<T>(
+        val value: T,
+        val code: Response.Code = Response.Code.Ok
+    ) : McuMgrResult<T>()
+
+    data class Error<T>(val code: Response.Code) : McuMgrResult<T>()
+
+    data class Failure<T>(val throwable: Throwable) : McuMgrResult<T>()
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:17]: It's not immediately clear what the difference is between an error vs. a failure?

Is there a need for the distinction? How would a consumer handle these differently?

File: core/src/main/kotlin/McuMgrResult.kt [hunk 17]
```diff
@@ -0,0 +1,122 @@
+package com.juul.mcumgr
+
+import com.juul.mcumgr.McuMgrResult.Error
+import com.juul.mcumgr.McuMgrResult.Failure
+import com.juul.mcumgr.McuMgrResult.Success
+import com.juul.mcumgr.message.Response
+
+sealed class McuMgrResult<out T> {
```
Comment (twyatt) [inline core/src/main/kotlin/McuMgrResult.kt:8]: This result class doesn't feel domain-specific. It looks like just a general result class.
That's not necessarily a problem, but it may be unnecessarily adding a lot of complexity as opposed to just propagating exceptions.

If we're just wrapping exceptions for the purpose of wrapping in a general result object, consumers of the library could do that with `runCatching`. By wrapping in a result you've added overhead and consumers have to unwrap if they want the exception. Seems more optimal to give the choice of creating the wrapped objects?

Now if this were more domain-specific (i.e. some of the sealed sub classes represented result states, such as upload failed due to timeout, or specific responses from the remote device, then suddenly this result is very informative/valuable).

File: core/src/main/kotlin/transfer/Uploader.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/main/kotlin/transfer/Uploader.kt:50]: I see. So the remote device can concurrently process these chunks while more chunks are coming in? And processing takes some time, hence why we're able to send multiple chunks before the remote has replied.

In other words, sending chunks sequentially would be substantially slower?

How much faster is it to send multiple chunks concurrently?

[the remark this answers is not in the document]
Comment (bgiori) [reply core/src/main/kotlin/transfer/Uploader.kt:50]: > I see. So the remote device can concurrently process these chunks while more chunks are coming in? And processing takes some time, hence why we're able to send multiple chunks before the remote has replied?

Exactly.

> How much faster is it to send multiple chunks concurrently?

Even using a capacity of 2 (two concurrent chunks) speeds up the upload by a factor of 2. As the capacity is increased the speedup flattens out, but e.g. capacity = 5 makes the image upload (even with our security protocols) speed up to less than 50 seconds (from 120 seconds).

File: core/src/main/kotlin/McuMgrResult.kt (hunk 16 above)
Comment (bgiori) [reply core/src/main/kotlin/McuMgrResult.kt:17]: The difference is an error response vs a total failure. E.g. a response with one field `{"rc":5}` would be an error response, while a ble disconnect would be a failure.

The reason to make the distinction is because, like HTTP requests, sometimes we expect a response to return an error. E.g. when downloading a core, the device may respond with `Response.Code.NoEntry` effectively telling us that there is no core available to download. This is not a failure, nor does it conform to the expected download response format, hence being a separate entity.

For most callers, the distinction is moot. However, it would be incorrect to lump errors and failures together, since doing so may cause failures which are not truly a failure.

File: core/src/main/kotlin/McuMgrResult.kt (hunk 17 above)
Comment (bgiori) [reply core/src/main/kotlin/McuMgrResult.kt:8]: Good point. I feel somewhat uncomfortable making public APIs so throw heavy, but it does provide a broader API for consumers to use.

Domain specific results might be better for wrapping classes (e.g. Downloader, Uploader) which perform additional logic on top of the base mcumgr command.

I'll keep McuMgrResult around to use for now, but I wont use it in the public API. If at the end,  McuMgrResult is no longer useful, I can remove it.

File: core/src/main/kotlin/McuMgrResult.kt (hunk 16 above)
Comment (twyatt) [reply core/src/main/kotlin/McuMgrResult.kt:17]: Gotchya. It'd be good to name them (if possible) to be more specific.

For example, Retrofit will throw either `HttpException` which indicates that an HTTP response was received (it was just not in the 2xx range) or an `IOException` to indicate a network failure. The naming makes it clear to the library consumer what they mean and how to react to them.

Failure vs. Error doesn't provide the same meaningful naming. If a more specific naming can't be found, then KDoc should be added to them to make it clear to the library consumer what their purpose is.

File: core/src/main/kotlin/McuMgrResult.kt (hunk 17 above)
Comment (twyatt) [reply core/src/main/kotlin/McuMgrResult.kt:8]: > I feel somewhat uncomfortable making public APIs so throw heavy

Agreed, and I definitely don't want to discourage use of result objects as they are more idiomatic (vs. throwing exceptions).

If possible, it would be preferred to return a domain-specific result object.

In this case, the result object just wasn't domain-specific. It looked like part of the design challenge is that this result object was wrapping your transport functions; which themselves don't provide a well defined possible error states (so we lean on the very general throw mechanics). Perhaps clearly defined failure states could help form a domain-specific result object for your transport functions. Then at the mcumgr layer (that calls the transport functions) you could easily map those results, or basically handle the ones than can be handled and then return the mcumgr-specific result object?

**Comment by philips77** [comment]
https://github.com/NordicSemiconductor/Android-nRF-Connect-Device-Manager
and
https://github.com/NordicSemiconductor/IOS-nRF-Connect-Device-Manager
